### PR TITLE
Submit validator votes

### PR DESCRIPTION
**Pull request by aeyakovenko**
* validators submit vote of the last PoH id, we dont really do anything with the vote yet, except update the liveness table.  this should help us with keep tracking of gossip table membership
* leaders process votes in the write stage
* @garious i need leaders to vote themselves

### CODE REVIEW

File: src/recorder.rs [hunk 1]
```diff
@@ -28,6 +32,18 @@ impl Recorder {
     }
 
     pub fn record(&mut self, transactions: Vec<Transaction>) -> Vec<Entry> {
+        for tx in transactions.iter() {
```
Comment (garious) [inline src/recorder.rs]: Clippy will have lots of feedback on this one.

**Comment by garious** [comment]
You went rogue again. Way too big of a WIP PR. Are you able to break this into separate PRs that you might expect us to grok in their entirety?

**Comment by aeyakovenko** [comment]
@garious its not ready.  i am testing this on a 96 cpu machine.  we can get 100 nodes to sync up in simulation (when the threads don't hang :))

let's sync up and see what can be split out and merged in pieces.  to add voting i had to touch a lot of pieces.

File: src/replicate_stage.rs [hunk 2]
```diff
@@ -1,57 +1,133 @@
...
+                let mut blob = shared_blob.write().unwrap();
+                let tx = Transaction::new_vote(&keypair, vote, last_id, 0);
+                let bytes = serialize(&tx)?;
+                let len = bytes.len();
+                blob.data[..len].copy_from_slice(&bytes);
+                blob.meta.set_addr(&addr);
+                blob.meta.size = len;
+            }
+            *last_vote = now;
+            vote_blob_sender.send(VecDeque::from(vec![shared_blob]))?;
```
Comment (garious) [inline src/replicate_stage.rs]: I'm curious, why are we sending a `VecDeque` on a channel?

File: src/transaction.rs [hunk 3]
```diff
@@ -59,6 +70,9 @@ pub enum Instruction {
     /// Tell the payment plan that the `NewContract` with `Signature` has been
     /// signed by the containing transaction's `PublicKey`.
     ApplySignature(Signature),
+
+    /// Vote for a PoH that is equal to the lastid of this transaction
```
Comment (garious) [inline src/transaction.rs]: Clever use of `last_id`

File: src/fullnode.rs [hunk 4]
```diff
@@ -37,11 +40,41 @@ pub enum OutFile {
     Path(String),
 }
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+/// Fullnode configuration to be stored in file
+pub struct Config {
+    pub network: ReplicatedData,
```
Comment (garious) [inline src/fullnode.rs]: Can you name that field whatever you're planning to rename`ReplicatedData` to? Maybe `NodeInfo`?

File: src/fullnode.rs [hunk 5]
```diff
@@ -37,11 +40,41 @@ pub enum OutFile {
...
+    pub fn new(bind_addr: &SocketAddr) -> Self {
+        let rnd = SystemRandom::new();
+        let pkcs8 = KeyPair::generate_pkcs8(&rnd)
+            .expect("generate_pkcs8 in mint pub fn new")
+            .to_vec();
+        let keypair =
+            KeyPair::from_pkcs8(Input::from(&pkcs8)).expect("from_pkcs8 in fullnode::Config new");
+        let pubkey = keypair.pubkey();
+        let network = ReplicatedData::new_leader_with_pubkey(pubkey, bind_addr);
+        Config {
```
Comment (garious) [inline src/fullnode.rs]:
```rust
Config { network, pkcs8 }
```

File: src/replicate_stage.rs (hunk 2 above)
Comment (aeyakovenko) [reply src/replicate_stage.rs]: Its a system call, so normally we would be sending a batch of things to do all in one go. just in this case there is only 1 item.

Comment (garious) [reply src/replicate_stage.rs]: But why a `VecDeque` and not a `Vec`?

Comment (aeyakovenko) [reply src/replicate_stage.rs]: i am not sure :)

File: src/voting.rs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+use entry::Entry;
+use hash::Hash;
+use signature::PublicKey;
+use transaction::{Instruction, Vote};
+
+pub fn entries_to_votes(entries: &Vec<Entry>) -> Vec<(PublicKey, Vote, Hash)> {
+    entries
+        .iter()
+        .flat_map(|entry| {
+            let vs: Vec<(PublicKey, Vote, Hash)> = entry
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/voting.rs]: i am not following, one that an external caller will `flat_map` with?  the voting right now deals with entries

File: src/replicate_stage.rs [hunk 7]
```diff
@@ -1,57 +1,133 @@
...
         let timer = Duration::new(1, 0);
-        let blobs = blob_receiver.recv_timeout(timer)?;
+        //coalesce all the available blobs into a single vote
+        let mut blobs = window_receiver.recv_timeout(timer)?;
+        while let Ok(mut more) = window_receiver.try_recv() {
+            blobs.append(&mut more);
+        }
         let blobs_len = blobs.len();
-        let entries = ledger::reconstruct_entries_from_blobs(blobs)?;
+        let entries = ledger::reconstruct_entries_from_blobs(blobs.clone())?;
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/replicate_stage.rs]: we pop the blobs to recycle them later, `reconstruct_entries_from_blobs` would need to take a recycler.

File: src/replicate_stage.rs [hunk 8]
```diff
@@ -1,57 +1,133 @@
...
+        if now - *last_vote > VOTE_TIMEOUT_MS {
+            let height = res?;
+            let last_id = bank.last_id();
+            let shared_blob = blob_recycler.allocate();
+            let (vote, addr) = {
+                let mut wcrdt = crdt.write().unwrap();
+                wcrdt.insert_votes(votes);
+                //TODO: doesn't seem like there is a synchronous call to get height and id
+                info!("replicate_stage {} {:?}", height, &last_id[..8]);
+                let (vote, addr) = wcrdt.new_vote(height, last_id)?;
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/replicate_stage.rs]: heh, cool `?` can basically go anywhere

File: src/replicate_stage.rs (hunk 2 above)
Comment (aeyakovenko) [reply src/replicate_stage.rs]: added  #572

**Pull request merged**